File: ruby/beer-song/beer_song.rb
# Beer song
class BeerSong
  NONE_LEFT = \
    "No more bottles of beer on the wall, no more bottles of beer.\n" \
    "Go to the store and buy some more, 99 bottles of beer on the wall.\n"

  ONE_LEFT = \
    "1 bottle of beer on the wall, 1 bottle of beer.\n" \
    "Take it down and pass it around, no more bottles of beer on the wall.\n"

  def initialize
  end

  def verse(n)
    if 0 == n
      NONE_LEFT
    elsif 1 == n
      ONE_LEFT
    else
      curr = pluralize(n, 'bottle')
      nv = pluralize(n - 1, 'bottle')
      "#{curr} of beer on the wall, #{curr} of beer.\n" \
        "Take one down and pass it around, #{nv} of beer on the wall.\n"
    end
  end

  def verses(first, last)
    # Would be better with .join("\n"), but tests specify 2 newlines after each
    first.downto(last).map { |x| verse(x) + "\n" }.join
  end

  def sing
    verses(99, 0)
  end

  private

  def pluralize(n, singular, plural = nil)
    if 1 == n
      "1 #{singular}"
    elsif plural
      "#{n} #{plural}"
    else
      "#{n} #{singular}s"
    end
  end
end

File: elixir/beer-song/beer_song.exs
defmodule BeerSong do
  @doc """
  Get a single verse of the beer song
  """
  @spec verse(integer) :: String.t
  def verse(number) do
    n = number - 1
    pluralized = n |> pluralize("bottle")
    next = n |> next |> pluralize("bottle")

    "#{ pluralized |> String.capitalize } of beer on the wall, #{ pluralized } of beer.\n#{ take(n) }, #{ next } of beer on the wall.\n"
  end

  @doc """
  Get the entire beer song for a given range of numbers of bottles.
  """
  @spec lyrics(Range.t) :: String.t
  def lyrics(range \\ 100..1) do
    range |> Enum.map(&verse/1) |> Enum.join("\n")
  end

  defp pluralize(0, item), do: "no more #{item}s"
  defp pluralize(1, item), do: "1 #{item}"
  defp pluralize(n, item), do: "#{n} #{item}s"

  defp next(0), do: 99
  defp next(n), do: n - 1

  defp take(0), do: "Go to the store and buy some more"
  defp take(1), do: "Take it down and pass it around"
  defp take(_), do: "Take one down and pass it around"
end
